### PR TITLE
Add dmenu mode, and new keybinding to print filename and quit

### PR DESCRIPTION
**Pull request by N-R-K**
Idea taken from https://github.com/muennich/sxiv/pull/405

I didn't think it would be good design to override a default keybinding to do something completely different depending on the mode. If someone presses `Enter` via muscle memory to switch mode it would end up selecting that img and quitting sxiv.

Having a separate binding (shift+enter) for printing to stdout and quitting seems more sane.

### CODE REVIEW

**Comment by N-R-K** [comment]
Btw, since these 2 commits are doing 2 different things, I think we shouldn't squash it.

File: options.c [hunk 1]
```diff
@@ -75,7 +76,7 @@ void parse_options(int argc, char **argv)
 	_options.clean_cache = false;
 	_options.private_mode = false;
 
-	while ((opt = getopt(argc, argv, "A:abce:fG:g:hin:N:opqrS:s:T:tvZz:")) != -1) {
+	while ((opt = getopt(argc, argv, "A:abce:fG:g:hin:N:oOpqrS:s:T:tvZz:")) != -1) {
```
Comment (XPhyro) [inline options.c:79]:
```suggestion
	while ((opt = getopt(argc, argv, "A:abce:fG:g:hin:N:OopqrS:s:T:tvZz:")) != -1) {
```

**Review by XPhyro — CHANGES_REQUESTED** [review]

**Comment by XPhyro** [comment]
This patch is easily achievable in `key-handler`, here's a minimal example:
```
#!/usr/bin/env sh

case "$1" in
    C-p|p) cat <&0;;
esac
```

I am not sure to include it in nsxiv. I'm also not strictly against it, so I'll leave the decision up to the others.

**Comment by N-R-K** [comment]
Two problems with the key-handler is that by default it takes more key-presses, and there's no visual indication of what you have selected. I have not used this, but scripts like this might be able to https://github.com/nagy135/subtube leverage the new dmenu mode.

With that being said, I'm not too sure about adding this in mainline nsxiv either. I'm perfectly fine having this as a community patch.

**Comment by eylles** [comment]
well, i can see the dmenu mode being useful, tho don't see much problem with this becoming part of mainline nsxiv as many scripts out there already are using sxiv as a sort of dmenu for images in a rather clunky way, if you ask me this is the kind of thing that would require a consensus of more people to decide if it is worth to add it to mainline or keep it as a community patch.

**Comment by N-R-K** [comment]
It seems that no one is totally in for merging this, but no one opposed either. I'm not sure what should be done in this situation.

IMO, the best course of action would be to have this as a community patch as of now. And then if there's enough interest in this we can revisit and reconsider adding it to mainline.

**Comment by eylles** [comment]
I think you took things with too much haste tho, should had waited for more opinions, remember the stable release was less than a day ago now.

**Comment by N-R-K** [comment]
If you believe so, then you may re-open it. And I'll wait more before closing other PRs in the future then, thanks for the feedback.

**Comment by eylles** [comment]
i honestly think this should remain open a little longer, and we should encourage more discussion about this feature and collect feedback before decifing, we already got xphyro's take that something to this function can be achieved in the keyhandler, so i added a similar option to mine
```sh
case "$1" in
    "e") echo "$file" ;;
esac
```
and i say it is a valid take, however i have been testing it and wondering if it ain't a bit too clunky compared to a more built in solution.
i say we should remember that not everyone can be always online and that with the case of adding features we have 2 cases, features that the faster they get merged/rejected the better and features that require time to evaluate, take a look, test and digest, and i think this is definitely the second case the PR ain't even a week old to really digest it and have everyone give their opinions after giving it a test ride and seeing if it does provide a real advantage or not over a solution in the keyhandler so in order to collect more feedback i'm reopening this i'd like for everyone to voice their opinions on this feature.

on the camp of having it as a keyhandler command, is there a way to kill nsxiv rom the key handler?

**Comment by TAAPArthur** [comment]
Putting aside whether this should go in or not, one thing I found interesting was the use of `ShiftMask` considering every entry uses either no modifier or `ControlMask`. 

Also for someone who is running sxiv without seeing its output, Shift-Enter to exit is very unintuitive.  Ctrl-Q or Esc would be better but that could be the vim talking.

**Comment by N-R-K** [comment]
> Shift-Enter to exit is very unintuitive

I think you're thinking about it the other way around, the main purpose here is not to _exit_ but rather to _pick_.

Currently you can just do `m -> q` which does the same, so having `shift+enter` is purely convenience. But one might argue, we're still using two keystroke, and I'm not sure what to do about that. As I said, I didn't think it would be good design to have `enter` do one thing on one mode and something radically different on another.

**Comment by TAAPArthur** [comment]
> > Shift-Enter to exit is very unintuitive
> 
> I think you're thinking about it the other way around, the main purpose here is not to _exit_ but rather to _pick_.

I wasn't clear. I meant for unintitive for someone not trying to use this behavior. For me if there is some behavior I think nsxiv has but don't know the keybinding, I'm going to take a guess and press some keys to see if I'm correct. If I try shift-enter and the program exists, I'm going to be surprised.

**Comment by eylles** [comment]
Interesting, i changed my previous example to:
```sh
case "$1" in
    "e") exec echo "$file" ;;
esac
```
And it still ain't killing nsxiv, i'm pretty sure this is killing the key handler script when it runs the echo command, however changing it to:

```sh
case "$1" in
    "e") 
        echo "$file"
        Killall nsxiv
        ;;
esac
```

Achieves the "desired" effect of dmenu emulation (kinda) so you can even do var=$(nsxiv -t /path/to/images/dir) and after running that keyhandler command the filenames will be saved to the var and nsxiv will close.

Will have to compare this keyhandler function to a branch with the patch for some days before i can really decide.

**Comment by TAAPArthur** [comment]
> And it still ain't killing nsxiv, i'm pretty sure this is killing the key handler script when it runs the echo command, however changing it to:

I believe if you use `kill $PPID` you can kill the parent nsxiv instead of all of them. I mentioned `exec` for a more general case where an a.out was spawned. Using `exec` allows such a program to just need to kill its parent instead of needing to kill its grandparent.

**Comment by N-R-K** [comment]
So one can use the keyhandler to not just emulate the marking behavior but also the pick_quit behavior. I see even less reasons to merge this then.

**Comment by eylles** [comment]
> So one can use the keyhandler to not just emulate the marking behavior but also the pick_quit behavior. I see even less reasons to merge this then.

it can be done but is kinda clunky.

**Comment by XPhyro** [comment]
Thinking out loud: I guess one benefit to including this in nsxiv would be that scripts can assume that this functionality exists, instead of relying on users patching their key-handler.

**Comment by N-R-K** [comment]
This patch shouldn't change how a script works.

**Comment by XPhyro** [comment]
From [this](https://github.com/nsxiv/nsxiv/pull/42#issuecomment-921167209) comment:
> [scripts like this](https://github.com/nagy135/subtube) might be able to leverage the new dmenu mode.

I meant that if this is included in nsxiv, scripts like this can directly use it instead of getting the user to edit their `key-handler`, or having to use the similar `-o` flag.

**Comment by N-R-K** [comment]
Yes, but the `-o` flag works fine. So scripts don't exactly need to change anything. Imo we should have it as a community patch for now.

- - -

And on that topic, we need to figure out how to host community patches.

**Comment by N-R-K** [comment]
I think enough time has passed and the opinion seems to be still the same, closing this. We can revisit this later if needed.

Also see: https://github.com/nsxiv/nsxiv-extra/blob/master/patches/dmenu-mode

**Pull request closed without merging**